lib/aws-manager.js

```diff
@@ -81,142 +82,164 @@ class AwsManager {
     let pricingStartDate = new Date();
     pricingStartDate.setMinutes(pricingStartDate.getMinutes() - 30);
 
-    let livingInstances;
-    let livingSpotRequests;
-    let deadInstances;
-    let deadSpotRequests = {};
-    let allPricingHistory;
-    let availableAZ;
-
-    try {
-      livingInstances = await this.ec2.describeInstances({
-        Filters: [
-          {
-            Name: 'key-name',
-            Values: [this.keyPrefix + '*'],
-          },
-          {
-            Name: 'instance-state-name',
-            Values: ['running', 'pending'],
-          },
-        ],
-      });
-      debug('fetched living instance state');
+    this.__previousApiState = this.__apiState;
+    let apiState = {
+      instances: [],
+      requests: [],
+    };
+    let deadState = {
+      instances: [],
+      requests: [],
+    };
 
-      livingSpotRequests = await this.ec2.describeSpotInstanceRequests({
-        Filters: [
-          {
-            Name: 'launch.key-name',
-            Values: [this.keyPrefix + '*'],
-          }, {
-            Name: 'state',
-            Values: ['open'],
-          },
-        ],
-      });
-      debug('fetched living spot request state');
+    let stalledKills = [];
+    let availableAZ = {};
+    let allPricingHistory = {};
+
+    // It would be better to do it by classification than by region for
+    // better concurrency, but this is easier and not too slow considering
+    // the 75s iteration frequency
+    _.forEach(this.ec2, async (obj, region) => {
+      let response = await Promise.all([
```

Comment:
I propose splitting this into 4 methods...
```js
  async _fetchLivingInstances(region) {
    // Get resevations
    let resevations = await this.ec2[region].describeInstances({
      Filters: [
        {
          Name: 'key-name',
          Values: [this.keyPrefix + '*'],
        },
        {
          Name: 'instance-state-name',
          Values: ['running', 'pending'],
        },
      ],
    }).promise().then(_.property('data.Reservations'))
    
    // For each instance add WorkerType and Region as properties
    return _.flatten(_.map(resevations, _.property('Instances'))).map(i => {
      i.WorkerType = this.parseKeyPairName(instance.KeyName).workerType;
      i.Region = region;
      return i;
    });
  }
```

Then instead of this long `_.forEach` I propose we make some simple `_.map` followed by `_.flatten`
```js
[
  apiState.instances,
  apiState.requests,
  ...
] = await Promise.all([
  Promise.all(_.map(this.ec2, region => this._fetchLivingInstances)).then(_.flatten),
  Promise.all(_.map(this.ec2, region => this._fetchLivingRequests)).then(_.flatten),
  ...
]);
```
Everything still runs in concurrently, but we've split it into smaller methods...
Each method returns a list, we avoid having to do the `apiState.instances.push(...)` by accepting that each method returns a list which we then `_.flatten` afterwards...

We then use unpacking from `await Promise.all` go  the result assigned to the correct `apiState` properties...

I realize that `availableAZ` and `allPricingHistory` won't fit this pattern nicely...
So we can either just hack them at the end... and disregard the return value when unpacking...
Or, probably better, we use `_.spread(_.assign)` instead of `_.flatten`...

This is all more functional, as in we returns values, and merge values into new values... as oppose to being imperative where we transform objects and arrays by adding things to them.

Generally, this functional-style programming will save you a lot pain and keep things simple.